__tests__/integration-metrics.test.js

```diff
@@ -1,145 +1,132 @@
+      responseBody += chunk;
+    }
+
+    expect(responseBody).not.to.equal('');
+    return JSON.parse(responseBody);
   }
-  expect(responseBody).not.to.equal('');
-  return JSON.parse(responseBody);
-}
 
-const randomApiKey = 'a-random-readme-api-key';
+  async function getPayload() {
+    if (process.env.HAS_HTTP_QUIRKS) {
+      return [sdkCall.req, sdkCall.body];
+    }
 
-describe('Metrics SDK Integration Tests', function () {
-  let metricsServer;
-  let httpServer;
-  let PORT;
+    const [req] = await once(server, 'request');
+    const body = await getBody(req);
+    return [req, body];
+  }
 
-  before(async function () {
-    metricsServer = http.createServer().listen(0, 'localhost');
-
-    await once(metricsServer, 'listening');
-    const { address, port } = metricsServer.address();
-    PORT = await getPort();
-
-    // In order to use child_process.spawn, we have to provide a
-    // command along with an array of arguments. So this is a very
-    // rudimental way of splitting the two values provided to us
-    // from the environment variable.
-    //
-    // I tried refactoring this to use child_process.exec, which just
-    // takes in a single string to run, but that creates it's own
-    // shell so we can't do `cp.kill()` on it later on (because that
-    // just kills the shell, not the actual command we're running).
-    //
-    // Annoyingly this works under macOS, so it must be a platform
-    // difference when running under docker/linux.
-    const [command, ...args] = process.env.EXAMPLE_SERVER.split(' ');
-    if (command === 'php') {
-      // Laravel's `artisan serve` command doesn't pick up `PORT` environmental variables, instead
-      // requiring that they're supplied as a command line argument.
-      args.push(`--port=${PORT}`);
-    }
+  beforeEach(function () {
+    sdkCall = {
+      req: {},
+      body: {},
+    };
+  });
 
-    httpServer = spawn(command, args, {
-      cwd: cwd(),
-      detached: true,
-      env: {
-        PORT,
-        METRICS_SERVER: new URL(`http://${address}:${port}`).toString(),
-        README_API_KEY: randomApiKey,
-        ...process.env,
-      },
+  before(async function () {
+    server = http
+      .createServer((req, res) => {
+        // Frameworks are funny. If we run this test suite with PHP we can access our Metrics
+        // request payload via a `request` event immediately on the HTTP server however if we run
+        // this same suite with Python, the `request` event sometimes gets emitted **after** we've
+        // already returned a response and closed the connection, resulting in our request payload
+        // being empty and the SDK in question crapping out with a connection read error.
+        //
+        // This quirk doesn't make sense and this logic here is extremely yikes but hey this test
+        // suite works now across all of our SDKs and is no longer flaky.
+        if (process.env.HAS_HTTP_QUIRKS) {
```

Comment:
Honestly no idea. Try removing the `HAS_HTTP_QUIRKS` env variable from either of the Python services in `docker-compose` and re-run it and see that all the tests fail.

Reply:
Weird... rather you than me to be the one to figure this out.

Reply:
![giphy](https://user-images.githubusercontent.com/33762/188676191-8246dc0d-d155-4b7c-b200-1b40b99c34cd.gif)